lib/src/workspace_store.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright 2025 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![expect(missing_docs)]
+
+use std::fmt::Debug;
+use std::fs;
+use std::io;
+use std::io::Write as _;
+use std::path::Path;
+use std::path::PathBuf;
+
+use jj_lib::file_util::IoResultExt as _;
+use jj_lib::file_util::PathError;
+use jj_lib::file_util::persist_temp_file;
+use jj_lib::lock::FileLock;
+use jj_lib::lock::FileLockError;
+use jj_lib::protos::workspace_store;
+use jj_lib::ref_name::WorkspaceNameBuf;
+use prost::Message as _;
+use tempfile::NamedTempFile;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum WorkspaceStoreError {
+    #[error("No such workspace: {0}")]
+    NoSuchWorkspace(String),
+    #[error("Failed to lock workspace store")]
+    Lock(#[source] FileLockError),
+    #[error(transparent)]
+    Io(#[from] io::Error),
+    #[error(transparent)]
+    Path(#[from] PathError),
+    #[error(transparent)]
+    ProstDecode(#[from] prost::DecodeError),
+}
+
+pub trait WorkspaceStore: Sized + Send + Sync + Debug {
+    fn name(&self) -> &str;
+
+    fn load(repo_path: &Path) -> Result<Self, WorkspaceStoreError>;
```

Comment:
Move `load()` to the implementation type? This isn't dyn compatible.
https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility